File: Graphs/Algorithmss/Centroid.py
import sys
from collections import deque
input = sys.stdin.readline

def read_tree():
    n = int(input())
    adj = [[] for _ in range(n+1)]
    for _ in range(n-1):
        a, b = map(int, input().split())
        adj[a].append(b)
        adj[b].append(a)
    return n, adj

def find_centroid(n, adj):
    parent = [0] * (n+1)
    order = [] 
    stack = [(1, 0)]
    while stack:
        u, p = stack.pop()
        parent[u] = p
        order.append(u)
        for v in adj[u]:
            if v == p: continue
            stack.append((v, u))
    dp = [1] * (n+1)
    for u in reversed(order):
        for v in adj[u]:
            if v == parent[u]: continue
            dp[u] += dp[v]

    u = 1
    limit = n // 2
    while True:
        moved = False
        for v in adj[u]:
            if v != parent[u] and dp[v] > limit:
                parent_cent = u
                u = v
                moved = True
                break
        if not moved:
            if n - dp[u] <= limit:
                return u
            u = parent[u]

n, adj = read_tree()
print(find_centroid(n, adj))

File: Graphs/Algorithmss/Centroid.cpp
#include <bits/stdc++.h>

using namespace std;
typedef long long ll;
typedef long double ld;
const ll INF = numeric_limits<ll>::max();

#define all(x) x.begin(), x.end()
#define rall(x) x.rbegin(), x.rend()
#define MOD 998244353
// #define MOD pow(10, 9)+7;
// #define MOD 3245284303;
#define pi acos(-1)
#define gcd __gcd
#define lcm(a, b) ((a) * (b) / gcd(a, b))
#define precision(n) cout << fixed << setprecision(n)
#define debug(arr) cerr << #arr << " = "; for (auto &x : arr) cerr << x << " "; cerr << endl;
#define fastio ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);

int main() {
    fastio;

    int n;
    cin >> n;
    vector<vector<int>> adj(n+1);
    for (int i = 0; i < n-1; i++) {
        int a, b;
        cin >> a >> b;
        adj[a].push_back(b);
        adj[b].push_back(a);
    }

    vector<int> parent(n+1, 0), order;
    order.reserve(n);
    stack<pair<int, int>> st;
    st.emplace(1, 0);

    while (!st.empty()) {
        auto [u, p] = st.top(); st.pop();
        parent[u] = p;
        order.push_back(u);
        for (int v : adj[u]) {
            if (v == p) continue;
            st.emplace(v, u);
        }
    }

    vector<int> dp(n+1, 1);
    for (int i = n-1; i >= 0; --i) {
        int u = order[i];
        for (int v : adj[u]) {
            if (v == parent[u]) continue;
            dp[u] += dp[v];
        }
    }

    int limit = n / 2;
    int u = 1;
    while (true) {
        bool moved = false;
        for (int v : adj[u]) {
            if (v != parent[u] && dp[v] > limit) {
                parent[u] = u;
                u = v;
                moved = true;
                break;
            }
        }
        if (moved) continue;

        if (n - dp[u] <= limit) {
            cout << u << '\n';
            break;
        }

        u = parent[u];
    }

    return 0;
}
